contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,253 @@
+# Out-of-tree client authentication providers
+
+Author: @ericchiang
+
+# Objective
+
+This document describes a credential rotation strategy for client-go using an exec-based
+plugin mechanism.
+
+# Motivation
+
+Kubernetes clients can provide three kinds of credentials: bearer tokens, TLS
+client certs, and basic authentication username and password. Kubeconfigs can either
+in-line the credential, load credentials from a file, or can use an `AuthProvider`
+to actively fetch and rotate credentials. `AuthProviders` are compiled into client-go
+and target specific providers (GCP, Keystone, Azure AD) or implement a specification
+supported but a subset of vendors (OpenID Connect).
+
+Long term, it's not practical to maintain custom code in kubectl for every provider. This
+is in-line with other efforts around kubernetes/kubernetes to move integration with cloud
+provider, or other non-standards-based systems, out of core in favor of extension points.
+
+Credential rotation tools have to be called on a regular basis in case the current
+credentials have expired, making [kubectl plugins](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/),
+kubectl's current extension point, unsuitable for credential rotation. It's easier
+to wrap `kubectl` so the tool is invoked on every command. For example, the following
+is a [real example](
+https://github.com/heptio/authenticator#4-set-up-kubectl-to-use-heptio-authenticator-for-aws-tokens)
+from Heptio's AWS authenticator:
+
+```
+kubectl --kubeconfig /path/to/kubeconfig --token "$(heptio-authenticator-aws token -i CLUSTER_ID -r ROLE_ARN)" [...]
+```
+
+Beside resulting in a long command, this potentially encourages distributions to
+wrap or fork kubectl, changing the way that users interact with different
+Kubernetes clusters.
+
+# Proposal
+
+This proposal builds off of earlier requests to [support exec-based plugins](
+https://github.com/kubernetes/kubernetes/issues/35530#issuecomment-256170024), and
+proposes that we should add this as a first-class feature of kubectl. Specifically,
```

Comment:
Will the implementation for this live in `k/k`, or mostly `k/client-go`, or be spread among different repos? I'd like to integrate as much of this as possible as transparently as possible in the cluster registry in order to move towards a `kubectl login` implementation with the cluster registry, and would appreciate being able to share code here without having to vendor `k/k`.

Reply:
This should live in `k/client-go`.